### DISCUSSION TITLE
Is it possible to parse from a stringstream without throwing ?

### DISCUSSION THREAD

**Discussion by DontBreakAlex** [nlohmann/json#2597] [Q&A] [answered]
I am receiving multiple JSONs on a tcp socket, here is my current implementation:

(This is in a method called each time the fd is read-ready)
```C++
char buff[WORKER_BUF_SZ];
int ret = ::read(fd, buff, WORKER_BUF_SZ);
rBuffer.write(buff, ret);
auto pos = rBuffer.tellg();
bool isJson = json::accept(rBuffer);
rBuffer.seekg(pos, rBuffer.beg);
if (isJson) {
	json j;
	j << rBuffer;
	onJson(srv, std::move(j));
}
```
This could be greatly simplified by using operator<< without throwing
 ```C++
char buff[WORKER_BUF_SZ];
int ret = ::read(fd, buff, WORKER_BUF_SZ);
rBuffer.write(buff, ret);
json j;
j << json::settrhow(false) << rBuffer;
if (!j.is_discarded()) {
	onJson(srv, std::move(j));
}
// Maybe seek the stream back if the parser doesn't do it
```
Alternatively, if I could know of many bytes json::parse() or accept() consumed, I could maintain a char array myself and discard the x first bytes after parsing.

Thank you for reading through.

**Comment by nlohmann** [comment]
Did you have a look at https://json.nlohmann.me/features/parsing/parse_exceptions/?

**Comment by DontBreakAlex** [comment]
Yes, and I does not provide a way to disable exceptions through operator<<, only through json::parse, which I can't use because the parser is called in strict mode and I may have several jsons in the buffer.

I can't throw because I may receive the json byte by byte, and if I throw each time it's going to take forever.

I'd like to call the parser this way: 
```C++
detail::parser<basic_json>(input_adapter(i), nullptr, false).parse(false, result);
```
json::parse will not allow me to do so (strict is hard coded to true), so I'm calling the parser myself:

```C++
void Worker::onReadReady(Server& srv) {
	char buff[WORKER_BUF_SZ];
	int ret = read(fd, buff, WORKER_BUF_SZ);
	rBuffer.write(buff, ret);
	json j;
	nlohmann::detail::parser<nlohmann::basic_json<>>(nlohmann::detail::input_adapter(rBuffer), nullptr, false).parse(false, j);
	if (!j.is_discarded()) {
		onJson(srv, std::move(j));
	}
}
```

1 - Is there a better way ?
2 - Should I expect this interface to change ?

Thanks !

**Reply by nlohmann** [reply]
- There is no way to disable exceptions when using `operator<<`. You need to call `json::parse` for that.
- I'm not sure from what exact type you are parsing, but you wrote "stringstream" in the title, so here is code that should do what you want:

```cpp
#include <iostream>
#include <sstream>
#include "json.hpp"

using json = nlohmann::json;

int main()
{
    std::stringstream ss;
    ss << "{\"valid\": \"json\"}";
    
    json j1 = json::parse(ss, nullptr, false);
    std::cout << j1 << std::endl;
    
    ss << "invalid";

    json j2 = json::parse(ss, nullptr, false);
    std::cout << j2 << std::endl;
}
```

Output:

```
{"valid":"json"}
<discarded>
```

**Reply by DontBreakAlex** [reply]
Thanks, 
I failed to mention explicitly that there may be 2 jsons in the stringstream (rBuffer), so this will not work, because json::parse calls the parser in strict mode. 

https://github.com/nlohmann/json/blob/92fa1d9582284a2eb0db8cc8769c57e88ec4cdc3/include/nlohmann/json.hpp#L6663
https://github.com/nlohmann/json/blob/92fa1d9582284a2eb0db8cc8769c57e88ec4cdc3/include/nlohmann/detail/input/parser.hpp#L78

**Reply by nlohmann** [reply]
Now I understand. Yes, then calling

```cpp
nlohmann::detail::parser<nlohmann::basic_json<>>(nlohmann::detail::input_adapter(rBuffer), nullptr, false).parse(false, j);
```

is fine.

There are several issues to improve the input adapter interface in the future, and I cannot guarantee that the line above will remain stable as it relies on the `internal` namespace.

**Comment by DontBreakAlex** [comment, marked as answer]
Resolved: https://github.com/nlohmann/json/discussions/2597#discussioncomment-288689
